### ISSUE TITLE
feat: Embedding Model Proposal for Databend(By ChatGPT4)

### ISSUE THREAD

**Issue by BohuTANG** [datafuselabs/databend#10689]
**Summary**

## Introduction

An embedding model is a machine learning technique that maps high-dimensional data into a lower-dimensional vector space. It is used to represent complex data such as words, sentences, or documents in a continuous and fixed-length vector. 

This representation can capture semantic and syntactic relationships, enabling a wide range of applications, including natural language processing, recommendation systems, and anomaly detection.

## Obtaining Embedding Vectors with OpenAI API

To obtain embedding vectors using the OpenAI API, you can use OpenAI's pre-trained language models. 

Here's an example using Python:
```
import openai

openai.api_key = "your_openai_api_key"

def get_embedding(text):
    response = openai.Completion.create(
        engine="davinci-codex",
        prompt=f"Embed the following text: {text}",
        max_tokens=16,
        n=1,
        stop=None,
        temperature=0.5,
    )
    embedding = response.choices[0].text.strip()
    return embedding

text = "Databend warehouse"
embedding = get_embedding(text)
print(embedding)

```

## Storing Embedding Vectors in Databend

To store the embedding vectors returned by the OpenAI API in Databend, you can create a table with a column of `Array(Float)` type to hold the vectors. Assuming you have already connected to a Databend instance:
```
CREATE TABLE embeddings (
    id INT,
    text VARCHAR(255) NOT NULL,
    vector ARRAY(FLOAT) NOT NULL
);

```

## Computing the Distance Between Vectors in Databend

Databend can compute the distance between the query vector and stored vectors using a built-in function called vector_distance, which calculates the distance between two ARRAY(FLOAT) inputs. This function can be used directly in SQL queries.

However, when dealing with large-scale datasets and high-dimensional vectors, calculating the vector distance for every pair of vectors can become computationally expensive and slow. To address this challenge, we propose the following techniques:

* `Approximate Nearest Neighbors (ANN)`: Implement an ANN algorithm like Annoy, HNSW, or Faiss to speed up the search for the most similar vectors. ANN algorithms can significantly reduce query time by sacrificing a small amount of accuracy.
* `Dimensionality Reduction`: Use dimensionality reduction techniques like PCA or t-SNE to reduce the size of the vectors before storing them in Databend. Smaller vectors will require less computation for distance calculations.
* `Indexing`: Create an index on the embedding vectors to speed up distance calculations. This can be done using specialized indexing methods like KD-trees or Ball-trees, which are designed to work well with high-dimensional data.

## Challenges and Future Work

The primary challenge with implementing a fast vector_distance function is efficiently handling large-scale datasets and high-dimensional vectors. 
The techniques proposed in section 4, such as using approximate nearest neighbors algorithms, dimensionality reduction, and specialized indexing methods, can help alleviate this issue. Further research and experimentation will be needed to determine the best approach for Databend Warehouse to handle fast vector distance calculations for massive datasets and high-dimensional vectors.

### Insert sample data

```
INSERT INTO embeddings (text, vector) VALUES
(1, 'Databend warehouse', ARRAY[0.12, 0.34, -0.56, 0.78]),
(2, 'Data warehouse', ARRAY[-0.15, 0.37, 0.29, -0.22]);
```

### Query

```
WITH query_vector AS (
    SELECT ARRAY[0.11, 0.33, -0.55, 0.77] AS vector
)
SELECT id, text, vector_distance(vector, query_vector.vector) AS distance
FROM embeddings, query_vector
ORDER BY distance ASC
LIMIT 1;
```

**Comment by mokeyish** [comment]
[vector_distance: Similarity Metrics](https://milvus.io/docs/metric.md#floating)

**Comment by BohuTANG** [comment]
From openai doc:
https://platform.openai.com/docs/guides/embeddings/which-distance-function-should-i-use

```
We recommend [cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity). The choice of distance function typically doesn’t matter much.

OpenAI embeddings are normalized to length 1, which means that:

Cosine similarity can be computed slightly faster using just a dot product
Cosine similarity and Euclidean distance will result in the identical rankings
```

**Comment by thatcort** [comment]
When is the vector index feature expected to be complete?

**Comment by BohuTANG** [comment]
> When is the vector index feature expected to be complete?

Indeed, there is a PR already https://github.com/datafuselabs/databend/pull/11318

But still a lot of work needs to do.
From Databend users case, their data is not large, so we make this ticket to low priority.

**Comment by thatcort** [comment]
I'm considering Databend for querying over large data sets of text and vectors. Vector indexing would allow replacing the current vector DB and save a lot of money by using object storage. Would be great if you raised the priority of that feature!

**Comment by BohuTANG** [comment]
Thank you for your explanation. 
We are planning to schedule this feature for the Q4 release, as there are currently many high-priority tasks that need to be completed.

**Comment by thatcort** [comment]
Another library worth looking at for vector ann support is USearch: https://unum-cloud.github.io/usearch/
